File: android/app/src/main/java/com/otoniel/sonylivemonitor/ArwGpsWriter.kt
package com.otoniel.sonylivemonitor

import android.location.Location
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.roundToLong

/**
 * Inserta un GPS IFD estandar en un Sony ARW (TIFF) sin tocar los datos RAW.
 *
 * Se anade un IFD0 nuevo al final y el puntero de cuatro bytes de la cabecera se
 * cambia al final de la operacion. El IFD0 y todos los bloques originales quedan
 * byte por byte intactos. Debe usarse sobre una copia temporal del archivo.
 */
object ArwGpsWriter {
    private const val TIFF_MAGIC = 42
    private const val TAG_GPS_IFD = 0x8825
    private const val TYPE_BYTE = 1
    private const val TYPE_ASCII = 2
    private const val TYPE_LONG = 4
    private const val TYPE_RATIONAL = 5
    private const val MAX_IFD_ENTRIES = 4096

    private data class Entry(val tag: Int, val bytes: ByteArray)
    private data class Fix(val latitude: Double, val longitude: Double, val altitude: Double, val time: Long)

    fun write(file: File, location: Location) =
        write(file, Fix(location.latitude, location.longitude, location.altitude, location.time))

    /** Entrada independiente de Android para pruebas de compatibilidad con ARW reales. */
    fun write(file: File, latitude: Double, longitude: Double, altitude: Double, time: Long) =
        write(file, Fix(latitude, longitude, altitude, time))

    private fun write(file: File, fix: Fix) {
        require(fix.latitude in -90.0..90.0 && fix.longitude in -180.0..180.0)
        RandomAccessFile(file, "rw").use { io ->
            require(io.length() in 8..0xfffffff0L) { "Unsupported ARW size" }
            val marker = ByteArray(2).also { io.readFully(it) }
            val order = when (String(marker, Charsets.US_ASCII)) {
                "II" -> ByteOrder.LITTLE_ENDIAN
                "MM" -> ByteOrder.BIG_ENDIAN
                else -> throw IllegalArgumentException("Not a TIFF/ARW file")
            }
            require(readU16(io, order) == TIFF_MAGIC) { "Not a TIFF/ARW file" }
            val oldIfdOffset = readU32(io, order)
            require(oldIfdOffset in 8 until io.length() - 2) { "Invalid ARW IFD0" }
            io.seek(oldIfdOffset)
            val count = readU16(io, order)
            require(count in 1..MAX_IFD_ENTRIES) { "Unsupported ARW IFD0" }
            val tableEnd = oldIfdOffset + 2 + count * 12L + 4
            require(tableEnd <= io.length()) { "Truncated ARW IFD0" }
            val existing = ArrayList<Entry>(count + 1)
            repeat(count) {
                val raw = ByteArray(12).also { io.readFully(it) }
                val tag = ByteBuffer.wrap(raw, 0, 2).order(order).short.toInt() and 0xffff
                if (tag != TAG_GPS_IFD) existing += Entry(tag, raw)
            }
            val nextIfd = ByteArray(4).also { io.readFully(it) }

            var appendAt = align2(io.length())
            require(appendAt < 0xfffffff0L) { "ARW is too large to extend" }
            if (appendAt != io.length()) { io.seek(io.length()); io.write(0) }
            val newIfdOffset = appendAt
            val entries = existing + Entry(TAG_GPS_IFD, ByteArray(12))
            val gpsIfdOffset = newIfdOffset + 2 + entries.size * 12L + 4
            val gps = buildGpsBlock(gpsIfdOffset, fix, order)

            io.seek(newIfdOffset)
            writeU16(io, entries.size, order)
            entries.sortedBy { it.tag }.forEach { entry ->
                if (entry.tag == TAG_GPS_IFD) {
                    writeU16(io, TAG_GPS_IFD, order)
                    writeU16(io, TYPE_LONG, order)
                    writeU32(io, 1, order)
                    writeU32(io, gpsIfdOffset, order)
                } else {
                    io.write(entry.bytes)
                }
            }
            io.write(nextIfd)
            io.write(gps)
            io.fd.sync()

            // Commit atomico dentro de la copia temporal: hasta aqui el ARW
            // original seguia siendo legible mediante su IFD0 anterior.
            io.seek(4)
            writeU32(io, newIfdOffset, order)
            io.fd.sync()
        }
    }

    private fun buildGpsBlock(offset: Long, location: Fix, order: ByteOrder): ByteArray {
        val entryCount = 10
        val dataStart = offset + 2 + entryCount * 12L + 4
        val extras = ArrayList<ByteArray>()
        var dataOffset = dataStart
        val entries = ArrayList<ByteArray>(entryCount)

        fun inline(tag: Int, type: Int, count: Long, value: ByteArray) {
            require(value.size <= 4)
            entries += entry(tag, type, count, value.copyOf(4), order)
        }
        fun external(tag: Int, type: Int, count: Long, value: ByteArray) {
            entries += entry(tag, type, count, u32Bytes(dataOffset, order), order)
            extras += value
            dataOffset += value.size
            if (dataOffset % 2L != 0L) { extras += byteArrayOf(0); dataOffset++ }
        }

        inline(0x0000, TYPE_BYTE, 4, byteArrayOf(2, 3, 0, 0))
        inline(0x0001, TYPE_ASCII, 2, byteArrayOf(if (location.latitude < 0) 'S'.code.toByte() else 'N'.code.toByte(), 0))
        external(0x0002, TYPE_RATIONAL, 3, dms(abs(location.latitude), order))
        inline(0x0003, TYPE_ASCII, 2, byteArrayOf(if (location.longitude < 0) 'W'.code.toByte() else 'E'.code.toByte(), 0))
        external(0x0004, TYPE_RATIONAL, 3, dms(abs(location.longitude), order))
        inline(0x0005, TYPE_BYTE, 1, byteArrayOf(if (location.altitude < 0) 1 else 0))
        external(0x0006, TYPE_RATIONAL, 1, rational(abs(location.altitude), 1000, order))
        val instant = Instant.ofEpochMilli(location.time)
        val utc = instant.atZone(ZoneOffset.UTC)
        external(0x0007, TYPE_RATIONAL, 3,
            rational(utc.hour.toDouble(), 1, order) + rational(utc.minute.toDouble(), 1, order) +
                rational(utc.second + utc.nano / 1_000_000_000.0, 1000, order))
        external(0x0012, TYPE_ASCII, 7, "WGS-84\u0000".toByteArray(Charsets.US_ASCII))
        external(0x001d, TYPE_ASCII, 11,
            (DateTimeFormatter.ofPattern("yyyy:MM:dd").format(utc) + "\u0000").toByteArray(Charsets.US_ASCII))

        return ByteBuffer.allocate((2 + entryCount * 12 + 4) + extras.sumOf { it.size }).order(order).apply {
            putShort(entryCount.toShort())
            entries.sortedWith(compareBy { ByteBuffer.wrap(it, 0, 2).order(order).short.toInt() and 0xffff }).forEach(::put)
            putInt(0)
            extras.forEach(::put)
        }.array()
    }

    private fun dms(value: Double, order: ByteOrder): ByteArray {
        val degrees = floor(value)
        val minutesValue = (value - degrees) * 60.0
        val minutes = floor(minutesValue)
        val seconds = (minutesValue - minutes) * 60.0
        return rational(degrees, 1, order) + rational(minutes, 1, order) + rational(seconds, 1_000_000, order)
    }

    private fun rational(value: Double, denominator: Int, order: ByteOrder): ByteArray =
        ByteBuffer.allocate(8).order(order).putInt((value * denominator).roundToLong().coerceIn(0, 0xffffffffL).toInt())
            .putInt(denominator).array()

    private fun entry(tag: Int, type: Int, count: Long, value: ByteArray, order: ByteOrder): ByteArray =
        ByteBuffer.allocate(12).order(order).putShort(tag.toShort()).putShort(type.toShort())
            .putInt(count.toInt()).put(value).array()

    private fun u32Bytes(value: Long, order: ByteOrder): ByteArray =
        ByteBuffer.allocate(4).order(order).putInt(value.toInt()).array()

    private fun align2(value: Long) = (value + 1L) and -2L
    private fun readU16(io: RandomAccessFile, order: ByteOrder): Int =
        ByteBuffer.wrap(ByteArray(2).also(io::readFully)).order(order).short.toInt() and 0xffff
    private fun readU32(io: RandomAccessFile, order: ByteOrder): Long =
        ByteBuffer.wrap(ByteArray(4).also(io::readFully)).order(order).int.toLong() and 0xffffffffL
    private fun writeU16(io: RandomAccessFile, value: Int, order: ByteOrder) = io.write(ByteBuffer.allocate(2).order(order).putShort(value.toShort()).array())
    private fun writeU32(io: RandomAccessFile, value: Long, order: ByteOrder) = io.write(u32Bytes(value, order))
}

File: ios/SonyLiveMonitor/ArwGpsWriter.swift
import Foundation

enum ArwGpsError: LocalizedError {
    case invalid(String)
    var errorDescription: String? {
        if case .invalid(let message) = self { return message }
        return "Invalid ARW"
    }
}

/// Anade un GPS IFD a un ARW/TIFF sin decodificar ni modificar los datos RAW.
enum ArwGpsWriter {
    private enum Endian { case little, big }
    private static let gpsTag: UInt16 = 0x8825

    static func write(_ fix: GeoFix, to url: URL) throws {
        var data = try Data(contentsOf: url, options: .mappedIfSafe)
        guard data.count >= 8 else { throw ArwGpsError.invalid("Truncated ARW") }
        let order: Endian
        if data[0] == 0x49 && data[1] == 0x49 { order = .little }
        else if data[0] == 0x4d && data[1] == 0x4d { order = .big }
        else { throw ArwGpsError.invalid("Not a TIFF/ARW file") }
        guard u16(data, 2, order) == 42 else { throw ArwGpsError.invalid("Not a TIFF/ARW file") }
        let oldIFD = Int(u32(data, 4, order))
        guard oldIFD >= 8, oldIFD + 2 <= data.count else { throw ArwGpsError.invalid("Invalid ARW IFD0") }
        let count = Int(u16(data, oldIFD, order))
        guard count > 0, count <= 4096, oldIFD + 2 + count * 12 + 4 <= data.count else {
            throw ArwGpsError.invalid("Unsupported ARW IFD0")
        }

        var entries: [(UInt16, Data)] = []
        for index in 0..<count {
            let start = oldIFD + 2 + index * 12
            let tag = u16(data, start, order)
            if tag != gpsTag { entries.append((tag, data.subdata(in: start..<(start + 12)))) }
        }
        let nextStart = oldIFD + 2 + count * 12
        let nextIFD = data.subdata(in: nextStart..<(nextStart + 4))
        if data.count % 2 != 0 { data.append(0) }
        let newIFD = data.count
        entries.append((gpsTag, Data()))
        let gpsOffset = newIFD + 2 + entries.count * 12 + 4
        let gps = gpsBlock(fix, at: gpsOffset, order)

        appendU16(&data, UInt16(entries.count), order)
        for entry in entries.sorted(by: { $0.0 < $1.0 }) {
            if entry.0 == gpsTag {
                appendU16(&data, gpsTag, order); appendU16(&data, 4, order)
                appendU32(&data, 1, order); appendU32(&data, UInt32(gpsOffset), order)
            } else {
                data.append(entry.1)
            }
        }
        data.append(nextIFD)
        data.append(gps)
        setU32(&data, 4, UInt32(newIFD), order)
        try data.write(to: url, options: .atomic)
    }

    private static func gpsBlock(_ fix: GeoFix, at offset: Int, _ order: Endian) -> Data {
        var entries: [(UInt16, Data)] = []
        var extras = Data()
        let entryCount = 10
        let dataStart = offset + 2 + entryCount * 12 + 4
        func entry(_ tag: UInt16, _ type: UInt16, _ count: UInt32, _ value: Data) -> Data {
            var out = Data(); appendU16(&out, tag, order); appendU16(&out, type, order)
            appendU32(&out, count, order); out.append(value); return out
        }
        func inline(_ tag: UInt16, _ type: UInt16, _ count: UInt32, _ value: Data) {
            var padded = value; while padded.count < 4 { padded.append(0) }
            entries.append((tag, entry(tag, type, count, padded)))
        }
        func external(_ tag: UInt16, _ type: UInt16, _ count: UInt32, _ value: Data) {
            var pointer = Data(); appendU32(&pointer, UInt32(dataStart + extras.count), order)
            entries.append((tag, entry(tag, type, count, pointer)))
            extras.append(value); if extras.count % 2 != 0 { extras.append(0) }
        }
        inline(0x0000, 1, 4, Data([2, 3, 0, 0]))
        inline(0x0001, 2, 2, Data([fix.latitude < 0 ? 0x53 : 0x4e, 0]))
        external(0x0002, 5, 3, dms(abs(fix.latitude), order))
        inline(0x0003, 2, 2, Data([fix.longitude < 0 ? 0x57 : 0x45, 0]))
        external(0x0004, 5, 3, dms(abs(fix.longitude), order))
        inline(0x0005, 1, 1, Data([fix.altitude < 0 ? 1 : 0]))
        external(0x0006, 5, 1, rational(abs(fix.altitude), denominator: 1000, order))
        let calendar = Calendar(identifier: .gregorian)
        let utc = TimeZone(secondsFromGMT: 0)!
        let parts = calendar.dateComponents(in: utc, from: fix.time)
        var clock = Data()
        clock.append(rational(Double(parts.hour ?? 0), denominator: 1, order))
        clock.append(rational(Double(parts.minute ?? 0), denominator: 1, order))
        clock.append(rational(Double(parts.second ?? 0), denominator: 1, order))
        external(0x0007, 5, 3, clock)
        external(0x0012, 2, 7, Data("WGS-84\0".utf8))
        let formatter = DateFormatter(); formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = utc; formatter.dateFormat = "yyyy:MM:dd"
        external(0x001d, 2, 11, Data((formatter.string(from: fix.time) + "\0").utf8))

        var out = Data(); appendU16(&out, UInt16(entryCount), order)
        entries.sorted(by: { $0.0 < $1.0 }).forEach { out.append($0.1) }
        appendU32(&out, 0, order); out.append(extras)
        return out
    }

    private static func dms(_ value: Double, _ order: Endian) -> Data {
        let degrees = floor(value), minutesValue = (value - degrees) * 60
        let minutes = floor(minutesValue), seconds = (minutesValue - minutes) * 60
        var out = Data(); out.append(rational(degrees, denominator: 1, order))
        out.append(rational(minutes, denominator: 1, order))
        out.append(rational(seconds, denominator: 1_000_000, order)); return out
    }
    private static func rational(_ value: Double, denominator: UInt32, _ order: Endian) -> Data {
        var out = Data(); appendU32(&out, UInt32(max(0, value * Double(denominator)).rounded()), order)
        appendU32(&out, denominator, order); return out
    }
    private static func u16(_ data: Data, _ offset: Int, _ order: Endian) -> UInt16 {
        let a = UInt16(data[offset]), b = UInt16(data[offset + 1]); return order == .little ? a | b << 8 : a << 8 | b
    }
    private static func u32(_ data: Data, _ offset: Int, _ order: Endian) -> UInt32 {
        let bytes = (0..<4).map { UInt32(data[offset + $0]) }
        return order == .little ? bytes[0] | bytes[1] << 8 | bytes[2] << 16 | bytes[3] << 24
                                : bytes[0] << 24 | bytes[1] << 16 | bytes[2] << 8 | bytes[3]
    }
    private static func appendU16(_ data: inout Data, _ value: UInt16, _ order: Endian) {
        if order == .little { data.append(UInt8(value & 0xff)); data.append(UInt8(value >> 8)) }
        else { data.append(UInt8(value >> 8)); data.append(UInt8(value & 0xff)) }
    }
    private static func appendU32(_ data: inout Data, _ value: UInt32, _ order: Endian) {
        let shifts = order == .little ? [0, 8, 16, 24] : [24, 16, 8, 0]
        shifts.forEach { data.append(UInt8((value >> UInt32($0)) & 0xff)) }
    }
    private static func setU32(_ data: inout Data, _ offset: Int, _ value: UInt32, _ order: Endian) {
        var bytes = Data(); appendU32(&bytes, value, order)
        data.replaceSubrange(offset..<(offset + 4), with: bytes)
    }
}
